Adds `build-jvm-vendors` command for building the JVM Vendors buildpack

dmikusa opened the pull request:
This command can be used to build a single JVM Vendors buildpack with all of the vendors, it can be used to build individual buildpacks one per vendor, or it can even be used to build combinations with a subset of JVM vendors.

It is designed to be run locally, so one can easily package up JVM vendor buildpacks manually, while also still being suitable for use in CI. When used in CI, it should be the same commands as run locally.

dmikusa committed:
Adds `build-jvm-vendors` command for building the JVM Vendors buildpack

This command can be used to build a single JVM Vendors buildpack with all of the vendors, it can be used to build individual buildpacks one per vendor, or it can even be used to build combinations with a subset of JVM vendors.

It is designed to be run locally, so one can easily package up JVM vendor buildpacks manually, while also still being suitable for use in CI. When used in CI, it should be the same commands as run locally.

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa committed:
Merge branch 'main' into jvm-builds

dmikusa committed:
Merge branch 'main' into jvm-builds

dmikusa committed:
Merge branch 'main' into jvm-builds

dmikusa committed:
Merge branch 'main' into jvm-builds

anthonydahanne commented:
oh wow, that's a cool feature!
Can you please tell me how to use it, after I `go build ./..` it ?

Thanks!

dmikusa commented:
If you check out this commit, you can `make`. It will build and put the binary into `$HOME/go/bin`. You can adjust that in the Makefile if you want it to go somewhere else. There's a variable at the top.

Then...

```
> libpak-tools build-jvm-vendors -h
Build JVM Vendors Buildpacks

Usage:
  libpak-tools build-jvm-vendors [flags]

Flags:
      --buildpack-id string             buildpack id (default "paketo-buildpacks/jvm-vendors")
      --buildpack-path string           path to buildpack directory
      --cache-location string           path to cache downloaded dependencies (default: $PWD/dependencies)
      --dependency-filter stringArray   one or more filters that are applied to exclude dependencies
  -h, --help                            help for build-jvm-vendors
      --include-all-vendors             include all of the vendors (default: false)
      --include-dependencies            whether to include dependencies (default: false)
      --publish                         publish the buildpack to a buildpack registry (default: false)
      --registry-name string            prefix for the registry to publish to (default: the buildpack id)
      --single-buildpack                build output is a single buildpack with listed vendors (default: false)
      --strict-filters                  require filter to match all data or just some data (default: false)
      --vendors stringArray             list of vendors to build
      --version stringArray             versions to substitute into buildpack.toml/extension.toml
```

There are a couple different things it'll let you do:

1. Build a single buildpack with one or more vendors
2. Build multiple buildpacks, one for each vendor.

The idea is CI can run this to generate the images, but a user could also run it if they want to build locally.

dmikusa committed:
Merge branch 'main' into jvm-builds

anthonydahanne committed:
chore: replace deprecated dep.SHA256 with dep.GetChecksum()

anthonydahanne commented:
here's the few things I tried:

```
libpak-tools build-jvm-vendors --registry-name anthonydahanne/libpacktesting  --version 100.99.0 --vendors paketo-buildpacks/bellsoft-liberica --buildpack-path ~/workspaces/paketo-buildpacks/bellsoft-liberica --publish
➜ Building multiple JVM Vendors buildpacks
  Building paketo-buildpacks/bellsoft-liberica@100.99.0
➜ Compile Buildpack

Paketo Buildpack for BellSoft Liberica 100.99.0
  https://github.com/paketo-buildpacks/bellsoft-liberica
  Creating package in /var/folders/_p/8q94q0px1tn6_b73xt67hvsw0000gn/T/BundleBuildpack848153362
  Pre-package with scripts/build.sh
    Adding buildpack.toml
    Adding linux/amd64/LICENSE
    Adding linux/amd64/NOTICE
    Adding linux/amd64/README.md
    Adding linux/amd64/bin/build
    Adding linux/amd64/bin/detect
    Adding linux/amd64/bin/helper
    Adding linux/amd64/bin/main
    Adding linux/arm64/LICENSE
    Adding linux/arm64/NOTICE
    Adding linux/arm64/README.md
    Adding linux/arm64/bin/build
    Adding linux/arm64/bin/detect
    Adding linux/arm64/bin/helper
    Adding linux/arm64/bin/main
➜ Package Buildpack: paketo-buildpacks/bellsoft-liberica
Successfully pushed manifest list anthonydahanne/libpacktesting to registry
Successfully published package anthonydahanne/libpacktesting and saved to registry
```

and when I tested it, I got:

```
pack build  --buildpack anthonydahanne/libpacktesting:latest   --buildpack paketo-buildpacks/java   pif
[...]
===> DETECTING
[detector] target distro name/version labels not found, reading /etc/os-release file
[detector] 10 of 26 buildpacks participating
[detector] paketo-buildpacks/bellsoft-liberica 100.99.0
[detector] paketo-buildpacks/ca-certificates   3.10.1
[detector] paketo-buildpacks/syft              2.12.0
[detector] paketo-buildpacks/maven             6.20.0
[detector] paketo-buildpacks/executable-jar    6.13.0
[...]

[builder] Paketo Buildpack for BellSoft Liberica 100.99.0
[builder]   https://github.com/paketo-buildpacks/bellsoft-liberica
[builder] panic: interface conversion: interface {} is []interface {}, not []map[string]interface {}
[builder]
[builder] goroutine 1 [running]:
[builder] github.com/paketo-buildpacks/libpak.NewBuildpackMetadata(0x4000254060)
[builder] 	/Users/anthony/go/pkg/mod/github.com/paketo-buildpacks/libpak@v1.73.0/buildpack.go:223 +0xa64
[builder] github.com/paketo-buildpacks/libpak.NewConfigurationResolver({{0x400023c262, 0x3}, {{0x400023c332, 0x23}, {0x400023c38c, 0x26}, {0x400023c412, 0x8}, {0x400023c2f2, 0x36}, ...}, ...}, ...)
[builder] 	/Users/anthony/go/pkg/mod/github.com/paketo-buildpacks/libpak@v1.73.0/buildpack.go:345 +0x2c
[builder] github.com/paketo-buildpacks/libjvm.Build.Build({{{{0x0, 0x0}, {0x482a78, 0x400010e038}}, {0x482d80, 0x400018b180}, {0x482d80, 0x400018b1c0}, {0x482d80, 0x400018b200}, ...}, ...}, ...)
[builder] 	/Users/anthony/go/pkg/mod/github.com/paketo-buildpacks/libjvm@v1.46.0/build.go:104 +0x27c
[builder] github.com/paketo-buildpacks/libpak.buildDelegate.Build({{_, _}}, {{{0x4000105796, 0xa}}, {{0x400023c262, 0x3}, {{0x400023c332, 0x23}, {0x400023c38c, 0x26}, ...}, ...}, ...})
[builder] 	/Users/anthony/go/pkg/mod/github.com/paketo-buildpacks/libpak@v1.73.0/build.go:42 +0x88
[builder] github.com/buildpacks/libcnb.Build({0x482cc0, 0x4000114fc0}, {0x4000125710, 0x4, 0x482ce0?})
[builder] 	/Users/anthony/go/pkg/mod/github.com/buildpacks/libcnb@v1.30.4/build.go:257 +0xfc0
[builder] github.com/buildpacks/libcnb.Main({0x482ca0, 0x4000114fb0}, {0x482cc0, 0x4000114fc0}, {0x4000125710, 0x4, 0x4})
[builder] 	/Users/anthony/go/pkg/mod/github.com/buildpacks/libcnb@v1.30.4/main.go:47 +0x1f0
[builder] github.com/paketo-buildpacks/libpak.Main({0x482a98?, 0x733220?}, {0x482ab8?, 0x4000158c88?}, {0x0, 0x0, 0x400010e038?})
[builder] 	/Users/anthony/go/pkg/mod/github.com/paketo-buildpacks/libpak@v1.73.0/main.go:28 +0x380
[builder] main.main()
[builder] 	/Users/anthony/workspaces/paketo-buildpacks/bellsoft-liberica/cmd/main/main.go:35 +0x1f8
[builder] ERROR: failed to build: exit status 2
ERROR: failed to build: executing lifecycle: failed with status code: 51
```

🙁

c0d1ngm0nk3y commented on builder/jvm_vendors.go:
```diff
@@ -0,0 +1,326 @@
+
+		if err := internal.UpdateTOMLFile(b.BuildpackTOMLPath, UpdateBuildpackConfiguration(map[string]interface{}{
+			"BP_JVM_VENDORS": strings.Join(vendorList, ","),
+			"BP_JVM_VENDOR":  "bellsoft-liberica",
```
Why is `bellsoft-liberica` always the default? If it is not part of `BP_JVM_VENDORS`, it will force the user to set `BP_JVM_VENDOR `, right?

dmikusa replied:
> What is the future plan - will there still be a sap-machine buildpack released? Created as a vendor specific buildpack via build-jvm-vendors.

Yes. We'll continue to release individual images per-vendor like we always have. I don't see any reason to stop doing that, as it would be a major change. The libpak-tools have been augmented with a build command for generating different permutations of buildpacks & vendors. We have yet to add it, but this buildpack will essentially have CI that runs once for each individual vendor buildpack and once to produce the buildpack with all of the vendors.

At some point, we'll change the composite Java buildpack to have the all-vendors buildpack so it's easier for users to switch between vendors.

> Why is bellsoft-liberica always the default? If it is not part of BP_JVM_VENDORS, it will force the user to set BP_JVM_VENDOR , right?

I think this is a fair question. Right now, it's purely historical. We're doing that here, cause that's what we have done in the past. I don't think there's really an issue of backward compatibility though since different JVM vendors should be interchangeable.

The one thing I do like about having Bellsoft as the default, which not a lot of other vendors have, is that they ship both a JVM and Native image build. There's some nice consistency there. I'm not sure that's a deal breaker though, since we have a different composite image for Java native image. We could in theory ship a different default there. Using jvm-vendors, we could bundle an image with just the vendors that support native image and plug-in a different default (default is hard code at the moment, but we could change that to be selected at the time you build the jvm-vendors based image).

c0d1ngm0nk3y reviewed (commented):
What is the future plan - will there still be a `sap-machine` buildpack released? Created as a vendor specific buildpack via `build-jvm-vendors`.

paketo-bot committed:
Bump Go Modules

Bumps Go modules used by the project. See the commit for details on what modules were updated.

Signed-off-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

nicolasbender committed:
Add support for RFC0059

Co-authored-by: Pavel Busko <pavel.busko@sap.com>

paketo-bot committed:
Bump Go Modules

Bumps Go modules used by the project. See the commit for details on what modules were updated.

Signed-off-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dmikusa committed:
Fixes bug where unnecessary files were copied into the buildpack

We need to copy buildpack.toml to the destination directory and run the package operation from the destination directory, otherwise we will copy things we don't want/expect into the buildpack image. This was causing images to be larger than necessary.

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa committed:
Update goreleaser format

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa committed:
Refactor the cli interface to be a little friendlier

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa committed:
Update Go 1.26 & latest pipeline-builder

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa committed:
Merge branch 'main' into jvm-builds

dmikusa committed:
Add test coverage & update README

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa committed:
Fix go lints

Signed-off-by: Daniel Mikusa <dan@mikusa.com>

dmikusa commented:
I finally got some time to test this more.

There are three modes of operation for this command:

1. Build a single buildpack with everything in it.
    `libpak-tools build-jvm-vendors --buildpack-id paketo-buildpacks/jvm-vendors@1.0.0 --single-buildpack --include-all-vendors`

2. Build a single buildpack with a subset of jvm vendors in it.
    `libpak-tools build-jvm-vendors --buildpack-id paketo-buildpacks/jvm-vendors@1.0.0 --single-buildpack --vendors azul-zulu --vendors oracle`

3. Build multiple buildpacks with a single jvm vendor in each (classic mode).
    `libpak-tools build-jvm-vendors --buildpack-id paketo-buildpacks/azul-zulu@1.0.0 --vendors azul-zulu --buildpack-id paketo-buildpacks/oracle@1.0.0 --vendors oracle`

All three tested and working with jvm-vendors buildpack!! 🎊

With this command, it is unsupported to...

- build multiple buildpacks with multiple jvm vendors
- use `include-all-buildpacks` when building multiple buildpacks (no way to know what versions to use)

For the most part, these limitations are due to the cli arguments and telling the command what to do. It's just not easy to tell it what to do for these situations. If a need arises to support one of these cases, we could be we'd need a different way to tell the command what to do, like a config file or something that can more easily represent the state of what we want it to build.
